examples/gcoap/gcoap_cli.c

```diff
@@ -230,6 +235,115 @@ static ssize_t _riot_board_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, vo
     }
 }
 
+static ssize_t _riot_board_handler_blocking(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx)
+{
+    (void)ctx;
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    coap_opt_add_format(pdu, COAP_FORMAT_TEXT);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+    xtimer_sleep(3);
+
+    /* write the RIOT board name in the response buffer */
+    if (pdu->payload_len >= strlen(RIOT_BOARD)) {
+        memcpy(pdu->payload, RIOT_BOARD, strlen(RIOT_BOARD));
+        return resp_len + strlen(RIOT_BOARD);
+    }
+    else {
+        puts("gcoap_cli: msg buffer too small");
+        return gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+}
+
+/* This is the information needed when the operation is known to take some time
+ * and a separate response is sent unconditionally. See the proxy application
+ * for an example of how to use a timer to send an empty ACK when the operation
+ * takes too long. */
+#define NONBLOCKING_READY (-1)
+static struct nonblocking_info {
+    /** A value of NONBLOCKING_READY indicates that none of this is initialized */
+    ssize_t token_length;
+    uint8_t token[GCOAP_TOKENLEN_MAX];
+    sock_udp_ep_t remote;
+    xtimer_t response_timer;
+} nonblocking = { .token_length = NONBLOCKING_READY };
+
+static void _riot_board_handler_flush(void *arg)
+{
+    struct nonblocking_info *nonblocking = (struct nonblocking_info *)arg;
+    uint8_t buf[64];
+
+    printf("Timer!\n");
+
+    coap_pkt_t pdu_backend;
+    /* Just to make copy-pasted code easier (and I trust the compiler
+     * simplifies this again) */
+    coap_pkt_t *pdu = &pdu_backend;
+
+    /* Can't do that as it won't expose the token; copy/pasting instead
+    gcoap_req_init(&pdu, &buf[0], sizeof(buf), COAP_CODE_CONTENT, NULL);
+    */
+    pdu->hdr = (coap_hdr_t *)buf;
+
+    /* generate token */
+    uint16_t msgid = 4; /* FIXME determined by random dice roll as we can't access gcoap internals */
```

Comment:
Would we need to expose the message id generation code in gcoap in order to fix this later? Then again, if we move the *separate response* generation into gcoap (if we decide to), then we do not need this line here anymore.

Reply:
Yes, and that's one of the open items. I don't expect gcoap to handle all of the separate response process as a framework, but we may instead of exposing the msgid-take-and-increment have a "prepare header with a new message ID"-ish function.